### PR TITLE
Allow setting of VM hardware version

### PR DESCRIPTION
**Pull request by gabriel-samfira**
This change adds a new model level flag, that allows operators to
set a newer compatibility version for the instances that get spawned
by juju.

The version set via model config must be of equal or higher version
than that of the source template, and it must be supported by the
remote compute resource it will be spawned on. Attempting to set a
version that does not conform to the mentioned rules, will be ignored,
and the VM will use the same HW compatibility version as the source
template.

## Checklist

 - [ ] Requires a [pylibjuju](https://github.com/juju/python-libjuju) change
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR
 - [ ] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed
 - [ ] Comments answer the question of why design decisions were made

## QA steps

```sh
juju model-config force-vm-hardware-version=12
juju add-machine
```
Check if the VM compatibility version is set to 12 in your vsphere deployment, for the newly spawned VM.

## Documentation changes

No changes required to CLI or API. This change adds a single model config option.

### CODE REVIEW

**Comment by hpidcock** [comment]
[External PR - Allow setting of VM hardware version](https://trello.com/c/6wO6P5Kg/1473-external-pr-allow-setting-of-vm-hardware-version)

**Comment by hpidcock** [comment]
@gabriel-samfira is this still WIP?

**Comment by gabriel-samfira** [comment]
@hpidcock Apologies for the delay. For some reason, not getting emails on new comments. I have added tests to this PR. Please take a look when time permits.

**Comment by hpidcock** [comment]
!!build!!

**Review by hpidcock — APPROVED** [review]
I'm happy with this unless @hmlanigan has any feedback

**Comment by hpidcock** [comment]
@hmlanigan do you think we need to extend the provider specific documentation to provide reference to these options https://juju.is/docs/vsphere-cloud

File: provider/vsphere/internal/vsphereclient/client.go [hunk 1]
```diff
@@ -538,6 +540,108 @@ func megabytesToKiB(diskMiB uint64) uint64 {
...
+	if cr == nil || cr.EnvironmentBrowser == nil {
+		return 0, errors.Errorf("invalid compute resource")
+	}
+	ref := cr.EnvironmentBrowser.Reference()
+	req := types.QueryConfigOption{
+		This: ref,
+	}
+	opt, err := methods.QueryConfigOption(ctx, c.client, &req)
+	if err != nil {
+		return 0, err
```
Comment (SimonRichardson) [inline provider/vsphere/internal/vsphereclient/client.go:553]:
```suggestion
		return 0, errors.Trace(err)
```

File: provider/vsphere/internal/vsphereclient/client.go [hunk 2]
```diff
@@ -538,6 +540,108 @@ func megabytesToKiB(diskMiB uint64) uint64 {
...
+}
+
+func (c *Client) getVMHardwareVersion(ctx context.Context, srcVM *object.VirtualMachine) (int64, error) {
+	if srcVM == nil {
+		return 0, errors.Errorf("source VM may not be nil")
+	}
+	var templateVM mo.VirtualMachine
+	err := srcVM.Properties(ctx, srcVM.Reference(), []string{"config.version"}, &templateVM)
+	if err != nil {
+		return 0, err
```
Comment (SimonRichardson) [inline provider/vsphere/internal/vsphereclient/client.go:574]:
```suggestion
		return 0, errors.Trace(err)
```

File: provider/vsphere/internal/vsphereclient/client.go [hunk 3]
```diff
@@ -538,6 +540,108 @@ func megabytesToKiB(diskMiB uint64) uint64 {
...
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	return parsed, nil
+}
+
+func (c *Client) parseVMXVersion(version string) (int64, error) {
+	fields := strings.Split(version, "-")
+	if len(fields) != 2 {
+		return 0, fmt.Errorf("invalid VMX version: %s", version)
```
Comment (SimonRichardson) [inline provider/vsphere/internal/vsphereclient/client.go:590]: Can we use `errors.Errorf`? Using the `errors` package will give us a stack trace

**Comment by hmlanigan** [comment]
@hpidcock, yes, I'll update the doc once 2.9 is out.  Good catch.  We also need to fix this bug: https://bugs.launchpad.net/juju/+bug/1884794

**Review by hmlanigan — COMMENTED** [review]
QA:

If I use a hw ver higher than the max, the machine fails with a nice error message.  However retry-provisioning doesn't work, the machine-status is "provisioning error". 

```console
Machine  State    DNS            Inst id        Series  AZ  Message
...
2        down                    pending        focal       hardware version 19 not supported by target (max version 17)
```

There seems to some disagreement about in which cases retry-provisioning should work: https://bugs.launchpad.net/juju/+bug/1843134/comments/2 and a few bugs on the topic.

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**